content/blog/using-lua-as-esp-idf-component-with-esp32/index.md

```diff
@@ -0,0 +1,187 @@
+
+
+### Key Components
+
+#### Lua 5.4 ESP-IDF Component
+
+Lua 5.4 is included as a component in the ESP-IDF project.
+The component is configured to work seamlessly with the ESP32.
+The Lua interpreter can execute scripts and interact with ESP-IDF functionalities.
+
+### Filesystem Support (SPIFFS)
+
+The SPIFFS filesystem is used to store Lua scripts on the ESP32.
+Scripts can be easily added, updated, or removed without modifying the firmware.
+This approach allows developers to manage Lua scripts separately from the main application code.
+
+#### Memory Usage Insights
+
+The project logs memory usage at various stages to provide insights into how much memory is available, especially when Wi-Fi is enabled.
+Understanding memory consumption is crucial for developing stable applications on resource-constrained devices like the ESP32.
+
+## Understanding the Code
+
+Main Application (main/esp32-lua-example.c) initializes the SPIFFS filesystem.
+Optionally initializes Wi-Fi and logs memory usage.
+Runs Lua scripts from files using the run_lua_file function.
+Key Function: `run_lua_file`
+
+```c
+void run_lua_file(const char *file_name, const char *description) {
+    ESP_LOGI(TAG, "Running Lua script: %s", description);
+
+    lua_State *L = luaL_newstate();
+    if (L == NULL) {
+        ESP_LOGE(TAG, "Failed to create Lua state");
+        return;
+    }
+
+    luaL_openlibs(L);
+
+    // Set Lua module search path
+    if (luaL_dostring(L, "package.path = package.path .. ';./?.lua;/spiffs/?.lua'")) {
+        ESP_LOGE(TAG, "Failed to set package.path: %s", lua_tostring(L, -1));
+        lua_pop(L, 1);
+    }
+
+    // Load and execute the Lua script from file
+    if (luaL_dofile(L, file_name) != LUA_OK) {
+        ESP_LOGE(TAG, "Error executing Lua script '%s': %s", file_name, lua_tostring(L, -1));
+        lua_pop(L, 1);
+    }
+
+    lua_close(L);
+    ESP_LOGI(TAG, "Finished running Lua script: %s", description);
+}
+```
+
+Creates a new Lua state and opens standard libraries.
+Sets the package.path to include the filesystem path where Lua scripts are stored.
+Executes the Lua script from the specified file.
+
+### Filesystem Initialization
+
+Function: `init_spiffs`
+
+```c
+void init_spiffs(void) {
+    ESP_LOGI(TAG, "Initializing SPIFFS");
+
+    esp_vfs_spiffs_conf_t conf = {
+        .base_path = "/spiffs",
+        .partition_label = NULL,
+        .max_files = 5,
+        .format_if_mount_failed = true
+    };
+
+    esp_err_t ret = esp_vfs_spiffs_register(&conf);
+
+    if (ret != ESP_OK) {
+        ESP_LOGE(TAG, "Failed to mount or format filesystem");
+        return;
+    }
+
+    size_t total = 0, used = 0;
+    ret = esp_spiffs_info(NULL, &total, &used);
+    if (ret == ESP_OK) {
+        ESP_LOGI(TAG, "SPIFFS total: %d, used: %d", total, used);
+    } else {
+        ESP_LOGE(TAG, "Failed to get SPIFFS partition information");
+    }
+}
+```
+
+
+### Getting Started
+
+Clone the Repository
```

Comment:
Looks like these are steps to complete to get started.

Why not numbering them or formatting them as separate steps in some other ways?